cloudml-template/template/trainer/model.py

```diff
@@ -123,58 +159,68 @@ def create_estimator(config):
-        # output_layer = None
+        # Create output (logits) layer given the hidden layers (usually without applying any activation functions)
+        logits = None
 
-        # Specify the model output (i.e. predictions) given the output layer
-        predictions = None
+        return logits
 
-        # Specify the export output based on the predictions
-        export_outputs = {
-            'predictions': tf.estimator.export.PredictOutput(predictions)
-        }
+    def _train_op_fn(loss):
+        """Returns the op to optimize the loss."""
 
-        # Calculate loss based on output and labels
-        loss = None
+        # Update learning rate using exponential decay method
+        current_learning_rate = update_learning_rate()
 
         # Create Optimiser
         optimizer = tf.train.AdamOptimizer(
-            learning_rate=task.HYPER_PARAMS.learning_rate)
+            learning_rate=current_learning_rate)
 
         # Create training operation
         train_op = optimizer.minimize(
             loss=loss, global_step=tf.train.get_global_step())
 
-        # Specify additional evaluation metrics
-        eval_metric_ops = None
+        return train_op
+
+    def _model_fn(features, labels, mode):
+        """ model function for the custom estimator"""
+
+        # Create the model structure and compute the logits
+        logits = _inference(features)
+
+        # Create the model's head:
+        # - tf.contrib.estimator.regression_head
+        # - tf.contrib.estimator.binary_classification_head
+        # - tf.contrib.estimator.multi_lablel_head
+        # - tf.contrib.estimator.multi_head
+        head = None
+
+        return head.create_estimator_spec(
+            features,
+            mode,
+            logits,
+            labels=labels,
+            train_op_fn=_train_op_fn
+        )
+
+    print("creating a custom model...")
 
-        # Provide an estimator spec
-        estimator_spec = tf.estimator.EstimatorSpec(mode=mode,
-                                                    loss=loss,
-                                                    train_op=train_op,
-                                                    eval_metric_ops=eval_metric_ops,
-                                                    predictions=predictions,
-                                                    export_outputs=export_outputs
-                                                    )
-        return estimator_spec
+    estimator = tf.estimator.Estimator(model_fn=_model_fn, config=config)
 
-    print("creating a custom estimator")
+    estimator = tf.contrib.estimator.add_metrics(estimator, metric_fn)
 
-    return tf.estimator.Estimator(model_fn=_model_fn,
-                                  config=config)
+    return estimator
 
 
 # ***************************************************************************************
-# THIS IS A HELPER FUNCTION TO CREATE GET THE HIDDEN LAYER UNITS
+# HELPER FUNCTIONS USED FOR CONSTRUCTING THE MODELS
```

Comment:
Should the user change these helper functions to adapt the template for their specific use case?

Reply:
OK, the comment is now changed to:

YOU NEED NOT TO CHANGE THESE HELPER FUNCTIONS USED FOR CONSTRUCTING THE MODELS